balance_strategy.go

```diff
@@ -127,3 +153,884 @@ func balanceStrategyHashValue(vv ...string) uint32 {
+func filterAssignedPartitions(currentAssignment map[string][]topicPartitionAssignment, partition2AllPotentialConsumers map[topicPartitionAssignment][]string) map[string][]topicPartitionAssignment {
+	assignments := deepCopyAssignment(currentAssignment)
+	for memberID, partitions := range assignments {
+		// perform in-place filtering
+		i := 0
+		for _, partition := range partitions {
+			if _, exists := partition2AllPotentialConsumers[partition]; exists {
+				partitions[i] = partition
+				i++
+			}
+		}
+		assignments[memberID] = partitions[:i]
+	}
+	return assignments
+}
+
+func removeValueFromStringSlice(s []string, e string) []string {
+	for i, v := range s {
+		if v == e {
+			s = append(s[:i], s[i+1:]...)
+			break
+		}
+	}
+	return s
+}
+
+func removeIndexFromStringSlice(s []string, i int) []string {
+	if len(s) == 0 {
+		return s
+	}
+	return append(s[:i], s[i+1:]...)
+}
+
+func removeTopicPartitionFromMemberAssignments(assignments []topicPartitionAssignment, topic topicPartitionAssignment) []topicPartitionAssignment {
+	for i, assignment := range assignments {
+		if assignment == topic {
+			return append(assignments[:i], assignments[i+1:]...)
+		}
+	}
+	return assignments
+}
+
+func memberAssignmentsIncludeTopicPartition(assignments []topicPartitionAssignment, topic topicPartitionAssignment) bool {
+	for _, assignment := range assignments {
+		if assignment == topic {
+			return true
+		}
+	}
+	return false
+}
+
+func sortPartitions(currentAssignment map[string][]topicPartitionAssignment, partitionsWithADifferentPreviousAssignment map[topicPartitionAssignment]consumerGenerationPair, isFreshAssignment bool, partition2AllPotentialConsumers map[topicPartitionAssignment][]string, consumer2AllPotentialPartitions map[string][]topicPartitionAssignment) []topicPartitionAssignment {
+	sortedPartitions := make([]topicPartitionAssignment, 0)
+	if !isFreshAssignment && areSubscriptionsIdentical(partition2AllPotentialConsumers, consumer2AllPotentialPartitions) {
+		// if this is a reassignment and the subscriptions are identical (all consumers can consumer from all topics)
+		// then we just need to simply list partitions in a round robin fashion (from consumers with
+		// most assigned partitions to those with least)
+		assignments := filterAssignedPartitions(currentAssignment, partition2AllPotentialConsumers)
+
+		// sortedMemberIDs contains a descending-sorted list of consumers based on how many valid partitions are currently assigned to them
+		sortedMemberIDs := sortMemberIDsByPartitionAssignments(assignments)
+		for i := len(sortedMemberIDs)/2 - 1; i >= 0; i-- {
+			opp := len(sortedMemberIDs) - 1 - i
+			sortedMemberIDs[i], sortedMemberIDs[opp] = sortedMemberIDs[opp], sortedMemberIDs[i]
+		}
+		for {
+			// loop until no consumer-group members remain
+			if len(sortedMemberIDs) == 0 {
+				break
+			}
+			updatedMemberIDs := make([]string, 0)
+			for _, memberID := range sortedMemberIDs {
```

Comment:
I'm not sure this behavior is the same. The Java code loops on a heap here, continually polling the member with the most partitions remaining, or the lexicographically largest member id.

In Javaland, each loop removes one partition from the remaining partitions for a member, and then pushes that member back. Only if the member has no more partitions does it not get pushed back (missed optimization to not push back if there are no more members...).

Thus, if there were two members, M1 with 1 partition and M2 with 10, M2 would be used first every time. It has more partitions, and when it has one partition left, it has a higher member ID.

This code looks to use M2 once, then M1, and then M2 the remainder of the time (since there is nothing left in M1 and it is gone).